File: java/Hangman.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Hangman {

    public static void main(String[] args) {
        if (args.length == 0) {
            System.exit(-1);
        }
        hangman(args[0]);
    }

    public static void hangman(String word) {
        int wrong = 0;
        List<String> stages = new ArrayList<>();
        stages.add("");
        stages.add("---------");
        stages.add("|   |");
        stages.add("|   0");
        stages.add("|  /|/");
        stages.add("|   //");
        stages.add("|");
        stages.add("|");

        List<String> guessedChar = new ArrayList<>();
        char[] board = new char[word.length()];
        for (int i = 0; i < word.length(); i++) {
            board[i] = '_';
        }

        System.out.println("Welcome to Hangman");

        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("Predict one letter!");
            String guess = scanner.nextLine();

            if (guess.isEmpty()) {
                System.out.println("invalid input");
                wrong++;
                continue;
            }

            // char prediction
            if (guess.length() == 1) {
                if (guessedChar.contains(guess)) {
                    System.out.println("Already predicted");
                    wrong++;
                } else if (word.contains(guess)) {
                    System.out.println("Correct!");
                    guessedChar.add(guess);
                    for (int i = 0; i < word.length(); i++) {
                        if (guess.charAt(0) == word.charAt(i)) {
                            board[i] = guess.charAt(0);
                        }
                    }
                } else {
                    System.out.println("Failure!");
                    wrong++;
                }
            } 
            // word prediction
            else {
                if (!word.equals(guess)) {
                    System.out.println("Failure!");
                    wrong++;
                } else {
                    System.out.println("Correct!");
                    board = word.toCharArray();
                }
            }

            System.out.println(board);
            
            stages.remove(0);
            for (String s : stages) {
                System.out.println(s);
            }

            if (new String(board).indexOf('_') == -1) {
                System.out.println("You win! Correct answer is " + word);
                System.exit(0);
                break;
            }

            if (stages.isEmpty()) {
                System.out.println("You lose! The correct answer is " + word);
                System.exit(1);
                break;
            }
        }
    }
}

File: python/hangman.py
def hangman(word):#引数wordはプレイヤーに当てて欲しい単語
    wrong = 0#プレイヤーが何回間違えたかを数える変数
    stages = ["",#文字列のリスト,pythonがこのリストの文字列を一行づつ出力すると、絵が完成。
              "_______      ",
              "|            ",
              "|      |     ",
              "|      0     ",
              "|     /|\    ",
              "|     / \    ",
              "|            "
              ]
    rletters = list(word)#引数wordをリスト化
    board = ["_"] * len(word)#文字列のリストでプレイヤーに見せるヒント。len 関数は引数に指定したオブジェクトの長さや要素の数を取得します。
    win = False#プレイヤーの勝敗
    print("ハングマンへようこそ!")
    while wrong < len(stages) - 1:#stagesの要素数から1を引いているのは
        print("\n")#ゲームの見栄えを良くするため空行を出力
        msg = "1文字を予想してね! "
        char = input(msg)
        if char in rletters:#if 特定の値 in リストなど:
        #Trueの時（含まれている時）に実行する処理
            cind = rletters.index(char)
            board[cind] = char
            rletters[cind] = '$'
        else:
        #Falseの時（含まれていない時）に実行する処理
            wrong += 1

        print(" ".join(board))
        e = wrong + 1
        print("\n".join(stages[0:e]))
        if "_" not in board:
            print("あなたの勝ち!")
            print(" ".join(board))
            win = True
            break
    if not win:#win変数がFalseの時に実行
        print("\n".join(stages[0:wrong+1]))
        print("あなたの負け!正解は {}.".format(word))

hangman("cat")